filesystem/luks.ksy

```diff
@@ -11,45 +11,120 @@ meta:
   encoding: ASCII
   endian: be
 doc: |
-  Linux Unified Key Setup (LUKS) is a format specification for storing disk
+  Linux Unified Key Setup (LUKS2) is a format specification for storing disk
   encryption parameters and up to 8 user keys (which can unlock the master key).
 doc-ref: https://gitlab.com/cryptsetup/cryptsetup/wikis/LUKS-standard/on-disk-format.pdf
+         https://gitlab.com/cryptsetup/cryptsetup/-/blob/main/docs/on-disk-format-luks2.pdf
+
 seq:
-  - id: partition_header
-    type: partition_header
+  - id: primary_binary_header
+    type: binary_header
+  - id: primary_json_area
+    type: json_area
+    if: primary_binary_header.version2 == 0x02
+  - id: secondary_binary_header
+    type: binary_header
+    if: primary_binary_header.version2 == 0x02
+  - id: secondary_json_area
+    type: json_area
+    if: primary_binary_header.version2 == 0x02 
 types:
-  partition_header:
+  binary_header:
     seq:
-      - id: magic
-        contents: [0x4C, 0x55, 0x4B, 0x53, 0xBA, 0xBE]
-      - id: version
-        contents: [0x00, 0x01]
+      - id: magic1
+        type: str
+        size: 4
+        valid:
+          any-of:
+            - '"LUKS"'
+            - '"SKUL"'
+      - id: magic2
+        contents: [0xBA, 0xBE]
+      - id: version1
+        contents: [0x00]
+      - id: version2
+        type: u1
+        valid:
+          min: 0x01
+          max: 0x02
+      # LUKS1
       - id: cipher_name_specification
         type: str
         size: 32
+        if: version2 == 0x01
       - id: cipher_mode_specification
         type: str
         size: 32
+        if: version2 == 0x01
```

Comment:
You should extract this bunch of consecutive fields specific to LUKS1/LUKS2 to new types, which'll allow you to drop `if: version2 == {0x01,0x02}` on the individual fields - you will use an `if` only once when you include these types in the header.

This has all the benefits of better readability, maintainability and easier (and less error-prone) usage of generated parsers in users' applications.